python/lsst/ctrl/bps/parsl/sites/torque.py

```diff
@@ -0,0 +1,331 @@
+
+        Returns
+        -------
+        executor : `HighThroughputExecutor`
+            Executor for Torque jobs.
+        """
+        nodes = get_bps_config_value(
+            self.site,
+            "nodes",
+            int,
+            nodes,
+            required=True,
+        )
+        queue = get_bps_config_value(
+            self.site,
+            "queue",
+            str,
+            queue,
+        )
+        walltime = get_bps_config_value(
+            self.site,
+            "walltime",
+            str,
+            walltime,
+            required=True,
+        )
+        tasks_per_node = get_bps_config_value(
+            self.site,
+            "tasks_per_node",
+            int,
+            tasks_per_node,
+        )
+        worker_init = get_bps_config_value(
+            self.site,
+            "worker_init",
+            str,
+            walltime,
+        )
+        scheduler_options = get_bps_config_value(
+            self.site,
+            "scheduler_options",
+            str,
+            scheduler_options,
+        )
+
+        job_name = get_workflow_name(self.config)
+        if scheduler_options is None:
+            scheduler_options = ""
+        else:
+            scheduler_options += "\n"
+        scheduler_options += f"#PBS -N {job_name}\n"
+        if queue:
+            scheduler_options += f"#PBS -q {queue}\n"
+        if not isinstance(nodes, int) or nodes < 1:
+            nodes = 1
+        if not isinstance(tasks_per_node, int) or tasks_per_node < 1:
+            tasks_per_node = 1
+        if not isinstance(walltime, str) or len(walltime) < 1:
+            walltime = "48:00:00"
+        if worker_init is None:
+            worker_init = ""
+
+        return HighThroughputExecutor(
+            label,
+            provider=TorqueProviderI(
+                nodes_per_block=nodes,
+                tasks_per_node=tasks_per_node,
+                queue=queue,
+                walltime=walltime,
+                scheduler_options=scheduler_options,
+                worker_init=worker_init,
+                **(provider_options or {}),
+            ),
+            mem_per_worker=mem_per_worker,
+            address=self.get_address(),
+            **(executor_options or {}),
+        )
+
+    def get_executors(self) -> list[ParslExecutor]:
+        """Get a list of executors to be used in processing.
+
+        Each executor should have a unique ``label``.
+        """
+        return [self.make_executor("torque")]
+
+    def select_executor(self, job: "ParslJob") -> str:
+        """Get the ``label`` of the executor to use to execute a job.
+
+        Parameters
+        ----------
+        job : `ParslJob`
+            Job to be executed.
+
+        Returns
+        -------
+        label : `str`
+            Label of executor to use to execute ``job``.
+        """
+        return "torque"
+
+
+class TorqueProviderI(TorqueProvider):
```

Comment:
I'm interested in what this subclass is for - it looks like you're trying to add a tasks-per-node parameter which would usually end up launching multiple copies of the Parsl worker pool on one node (rather than having one process worker pool manage the whole node). Is this what you're intending / is this actually what happens?

Reply:
In the usual Parsl model, you'd run one copy of `process_worker_pool.py` on each node, and that worker pool would be in charge of running multiple tasks at once. The command line you specify has an option `-c 1.0` which means 1 core per worker.

So the worker pool code should run as many workers (and so, as many simultaneous tasks) as you have cores on your worker node: that is the code that is in charge of running multiple workers, not `mpirun`.

Have a look in your run directory (deep inside runinfo/....) for a file called `manager.log`. You should see one per node (or with your configuration above, 24 per node) and inside those files you should see a log line like this:

```
2024-09-13 12:54:44.837 parsl:254 72 MainThread [INFO]  Manager will spawn 8 workers
```

How many workers do you see there?

Reply:
you should see manager log in subdirectories near the interchange.log file like this:

```
$ find .
.
./block-0
./block-0/4b0b874bc188
./block-0/4b0b874bc188/worker_1.log
./block-0/4b0b874bc188/manager.log                   # <- HERE
./block-0/4b0b874bc188/worker_6.log
./block-0/4b0b874bc188/worker_7.log
./block-0/4b0b874bc188/worker_0.log
./block-0/4b0b874bc188/worker_3.log
./block-0/4b0b874bc188/worker_5.log
./block-0/4b0b874bc188/worker_2.log
./block-0/4b0b874bc188/worker_4.log
./certificates
./certificates/client.key
./certificates/server.key
./certificates/server.key_secret
./certificates/client.key_secret
./interchange.log
```

`manager.log` will only be created when a Parsl worker pool (`process_worker_pool.py`) start running on a worker node - if your batch job is still in the queue, that file won't exist yet.

Reply:
ok, so running 1 copy of the process worker pool on a node should give you 28 workers on that node (and so, 28 simultaneous tasks) - and running 12 copies should give you 12*28=336 workers on a node.

Maybe step back and describe what you are seeing that suggests there are not 28 workers with mainline Parsl, and we can debug that.

Reply:
> I have many mangers with different manager IDs

right, you asked for 24 managers to run (2 nodes x 12 managers per node) and so I'd expect to see these 24 here
.

Reply:
ok. are you running in an environment which allocations fractions of a node to a batch job? so that many batch jobs might be running at once on the same node?

Reply:
(rather than you being allocated an entire node, as happens on NERSC perlmutter?)

Reply:
first I want to understand what isn't working with normal Parsl in your execution environment -- the usual model (and all of the words used in configuration) expect that you can run one process worker pool per node, that code will see (in your case) 28 cores, and then 28 workers will be able to make use of all 28 cores.

I think you are saying that is not happening with normal parsl: 28 workers are running but they are not using all of the cores?

Reply:
each worker pool will try to spawn 28 workers because it can see 28 cores on the node (even if it is not allowed to use them all!) - separate from however many cores PBS has allocated for your batch job to use (12 per node)

In your setup, i think you should make each worker pool try to use only 1 core, so that when you run 12 worker pools per node, you get 1 x 12 = 12 workers on each node. Have a look at the max_workers Parsl configuration parameter - for example, see how it is configured at in2p3: https://github.com/lsst/ctrl_bps_parsl/blob/f08a03362232706d3236e1b07ae61ada8b63f416/python/lsst/ctrl/bps/parsl/sites/ccin2p3.py#L231

I will open an issue in Parsl for this, though: In the Parsl slurm provider, there is an option to request a whole node, but not in the PBS providers. This code in Parsl is getting confused because you want to tell PBS that you want 12 cores (or perhaps a whole node), but you want to only run 1 copy of the worker pool (that itself should run 12 workers)

Reply:
there are lots of different ways to change things to get what you want, so it is quite confusing.

You could try this:

i) set the number of nodes in your job to 1   (so if you want to run on multiple nodes, you launch multiple blocks/multiple batch jobs)

ii) use the change you have made in this PR to set task_per_node to 12 - so that 12 cores are requested in #PBS nodes=...

iii) use the `SimpleLauncher` instead of the `MpiRunLauncher` here:

https://github.com/lsst/ctrl_bps_parsl/pull/36/files#diff-e5ba88552b57b323bd184a741f622b7cc7b3a4090d5ac09456f7a8fe85fcc75cR287

so that only a single copy of the process worker pool is launched in each batch job - rather than using mpirun to launch many copies of it

iv) tell the process worker pool to use 12 workers per pool, using max_workers = 12.

That should result in batch jobs where each batch job:

* gets 12 cores from PBS
* runs one copy of the Parsl process worker pool
* the process worker pool runs 12 workers

Reply:
`SimpleLauncher` does not support running on two nodes.

The model I wrote above has 1 node per block/per batch job - and if you want to use two nodes, set the max_blocks paramter to 2. so that you get two separate batch jobs that look like this.

(I opened a Parsl issue https://github.com/Parsl/parsl/issues/3616 to request that the Parsl team try to make this interface nicer, some time in the future)